extensions/amp-delight-player/0.1/amp-delight-player.js

```diff
@@ -0,0 +1,700 @@
+    Services.videoManagerForDoc(this.element).register(this);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const src = this.baseURL_ + '/player/' + this.contentID_;
+    const iframe = createFrameFor(this, src);
+
+    iframe.setAttribute('allow', 'vr');
+
+    this.unlistenMessage_ = listen(this.win, 'message', event => {
+      this.handleDelightMessage_(event);
+    });
+
+    this.iframe_ = iframe;
+
+    this.registerEventHandlers_();
+
+    return this.loadPromise(iframe);
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+
+    const deferred = new Deferred();
+    this.playerReadyPromise_ = deferred.promise;
+    this.playerReadyResolver_ = deferred.resolve;
+
+    this.unregisterEventHandlers_();
+
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const placeholder = this.win.document.createElement('div');
+    const src = this.baseURL_ + '/poster/' + this.contentID_;
+    placeholder.setAttribute('placeholder', '');
+
+    setStyle(placeholder, 'background-image', 'url(' + src + ')');
+    setStyle(placeholder, 'background-repeat', 'no-repeat');
+    setStyle(placeholder, 'background-size', 'cover');
+    setStyle(placeholder, 'background-position', '50%');
+    setStyle(placeholder, 'width', '100%');
+    setStyle(placeholder, 'height', '100%');
+
+    this.placeholderEl_ = placeholder;
+
+    return placeholder;
+  }
+
+  /** @override */
+  firstLayoutCompleted() {
+    const el = this.placeholderEl_;
+    const isInViewport = this.isInViewport();
+
+    if (el && isInViewport) {
+      return new Promise(resolve => {
+
+        const onTransitionEnd = () => {
+          el.removeEventListener('transitionend', onTransitionEnd, false);
+          resolve();
+        };
+
+        el.classList.add('faded');
+        el.addEventListener('transitionend', onTransitionEnd, false);
+      }).then(() => super.firstLayoutCompleted());
+    } else {
+      return super.firstLayoutCompleted();
+    }
+  }
+
+  /** @override  */
+  pauseCallback() {
+    if (this.iframe_ && this.iframe_.contentWindow) {
+      this.pause();
+    }
+    return super.pauseCallback();
+  }
+
+  /** @override */
+  resumeCallback() {
+    if (this.iframe_ && this.iframe_.contentWindow) {
+      this.play(false);
+    }
+    return super.resumeCallback();
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  handleDelightMessage_(event) {
+    if (event.source !== this.iframe_.contentWindow) {
+      return;
+    }
+
+    const data = objOrParseJson(getData(event));
+    if (data === undefined || data['type'] === undefined) {
+      return; // We only process valid JSON.
+    }
+
+    const {element} = this;
+
+    switch (data['type']) {
```

Comment:
This switch statement is very table-like. Consider using a more general approach. As it stands it's a little bit hard to read and maintain.

Reply:
The switch cases are not all of equal structure. Only 5 out of 14 could be generalized. Is having a mix between custom action on some events and a default fallback action for the other events advisable and better to maintain/understand? What would you suggest as a solution?